### ISSUE TITLE
When running `xclip` in a function, it cannot return to the main script.

### ISSUE THREAD

**Issue by hongyi-zhao** [astrand/xclip#134]
On Ubuntu 20.04.3 LTS, I check the self-compiled git master version of [xclip](https://github.com/astrand/xclip) with the following bash script:

```bash
#! /bin/bash
myclip() {
xclip -selection clipbord -in <<< "hello world"
echo x
return
}
set -x
while true; do
y="$(myclip)"
done
```

It will be blocked there forever, as shown below. In other words, when running `xclip` in a function, it cannot return to the main script. 

![image](https://user-images.githubusercontent.com/11155854/161712904-990d841b-b333-412e-893c-62f75f238199.png)

Please refer to the related discussion [here](https://github.com/mviereck/x11docker/issues/428#issuecomment-1088423451).

**Comment by mviereck** [comment]
The blocking only happens if the output of the function is captured with `$(...)`.
It does not happen if the function is called without `$(...)` in the main loop:
```
#! /bin/bash
myclip() {
  xclip -selection clipbord -in <<< "hello world" 
  echo x
  return
}
set -x
while true; do
  myclip
done
```
However, I don't understand why the first example blocks.
I assume it is somehow related to xclip forking itself into background. 
I have tried with `nohup xclip ... &`, but that does not help.

**Comment by hongyi-zhao** [comment]
See the following debug output with strace:

```bash
$ cat myclip.sh 
#! /bin/bash
myclip() {
xclip -selection clipbord -in <<< "hello world"
echo x
return
}

while true; do
y="$(myclip)"
done

$ strace bash myclip.sh 
execve("/usr/bin/bash", ["bash", "myclip.sh"], 0x7ffd06c7d528 /* 102 vars */) = 0
brk(NULL)                               = 0x563292f6f000
arch_prctl(0x3001 /* ARCH_??? */, 0x7ffe21051040) = -1 EINVAL (Invalid argument)
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/tls/haswell/x86_64/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/tls/haswell/x86_64", 0x7ffe21050290) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/tls/haswell/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/tls/haswell", 0x7ffe21050290) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/tls/x86_64/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/tls/x86_64", 0x7ffe21050290) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/tls/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/tls", 0x7ffe21050290) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/haswell/x86_64/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/haswell/x86_64", 0x7ffe21050290) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/haswell/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/haswell", 0x7ffe21050290) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/x86_64/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/x86_64", 0x7ffe21050290) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
openat(AT_FDCWD, "tls/haswell/x86_64/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/haswell/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/x86_64/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "haswell/x86_64/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "haswell/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "x86_64/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "libtinfo.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=203413, ...}) = 0
mmap(NULL, 203413, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e02e000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libtinfo.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\240\346\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=192032, ...}) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x14a80e02c000
mmap(NULL, 194944, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x14a80dffc000
mmap(0x14a80e00a000, 61440, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0xe000) = 0x14a80e00a000
mmap(0x14a80e019000, 57344, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1d000) = 0x14a80e019000
mmap(0x14a80e027000, 20480, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x2a000) = 0x14a80e027000
close(3)                                = 0
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/haswell/x86_64/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/haswell/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/x86_64/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "haswell/x86_64/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "haswell/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "x86_64/libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "libdl.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libdl.so.2", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0 \22\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=18816, ...}) = 0
mmap(NULL, 20752, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x14a80dff6000
mmap(0x14a80dff7000, 8192, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1000) = 0x14a80dff7000
mmap(0x14a80dff9000, 4096, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3000) = 0x14a80dff9000
mmap(0x14a80dffa000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3000) = 0x14a80dffa000
close(3)                                = 0
openat(AT_FDCWD, "/home/werner/Public/repo/github.com/SCM-NV/ftl.git/install/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/haswell/x86_64/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/haswell/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/x86_64/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "haswell/x86_64/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "haswell/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "x86_64/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\360q\2\0\0\0\0\0"..., 832) = 832
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
pread64(3, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0", 32, 848) = 32
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\t\233\222%\274\260\320\31\331\326\10\204\276X>\263"..., 68, 880) = 68
fstat(3, {st_mode=S_IFREG|0755, st_size=2029224, ...}) = 0
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
pread64(3, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0", 32, 848) = 32
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\t\233\222%\274\260\320\31\331\326\10\204\276X>\263"..., 68, 880) = 68
mmap(NULL, 2036952, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x14a80de04000
mprotect(0x14a80de29000, 1847296, PROT_NONE) = 0
mmap(0x14a80de29000, 1540096, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x25000) = 0x14a80de29000
mmap(0x14a80dfa1000, 303104, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x19d000) = 0x14a80dfa1000
mmap(0x14a80dfec000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1e7000) = 0x14a80dfec000
mmap(0x14a80dff2000, 13528, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x14a80dff2000
close(3)                                = 0
mmap(NULL, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x14a80de01000
arch_prctl(ARCH_SET_FS, 0x14a80de01740) = 0
mprotect(0x14a80dfec000, 12288, PROT_READ) = 0
mprotect(0x14a80dffa000, 4096, PROT_READ) = 0
mprotect(0x14a80e027000, 16384, PROT_READ) = 0
mprotect(0x563292dda000, 16384, PROT_READ) = 0
mprotect(0x14a80e08d000, 4096, PROT_READ) = 0
munmap(0x14a80e02e000, 203413)          = 0
openat(AT_FDCWD, "/dev/tty", O_RDWR|O_NONBLOCK) = 3
close(3)                                = 0
brk(NULL)                               = 0x563292f6f000
brk(0x563292f90000)                     = 0x563292f90000
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=2996, ...}) = 0
read(3, "# Locale name alias data base.\n#"..., 4096) = 2996
read(3, "", 4096)                       = 0
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=352, ...}) = 0
mmap(NULL, 352, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e08c000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/gconv/gconv-modules.cache", O_RDONLY) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=27002, ...}) = 0
mmap(NULL, 27002, PROT_READ, MAP_SHARED, 3, 0) = 0x14a80e059000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_MEASUREMENT", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_MEASUREMENT", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=23, ...}) = 0
mmap(NULL, 23, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e058000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_TELEPHONE", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_TELEPHONE", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=51, ...}) = 0
mmap(NULL, 51, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e057000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_ADDRESS", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_ADDRESS", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=150, ...}) = 0
mmap(NULL, 150, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e056000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_NAME", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_NAME", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=62, ...}) = 0
mmap(NULL, 62, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e055000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_PAPER", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_PAPER", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=34, ...}) = 0
mmap(NULL, 34, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e054000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_MESSAGES", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_MESSAGES", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_MESSAGES/SYS_LC_MESSAGES", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=59, ...}) = 0
mmap(NULL, 59, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e053000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_MONETARY", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_MONETARY", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=286, ...}) = 0
mmap(NULL, 286, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e052000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_COLLATE", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_COLLATE", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=2586930, ...}) = 0
mmap(NULL, 2586930, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80db89000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_TIME", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_TIME", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=3268, ...}) = 0
mmap(NULL, 3268, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e051000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_NUMERIC", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_NUMERIC", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=54, ...}) = 0
mmap(NULL, 54, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e050000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/en_CA.UTF-8/LC_CTYPE", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/en_CA.utf8/LC_CTYPE", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=340640, ...}) = 0
mmap(NULL, 340640, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80db35000
close(3)                                = 0
getuid()                                = 1000
getgid()                                = 1000
geteuid()                               = 1000
getegid()                               = 1000
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
ioctl(-1, TIOCGPGRP, 0x7ffe21050e94)    = -1 EBADF (Bad file descriptor)
sysinfo({uptime=29231, loads=[228480, 221344, 235552], totalram=134946013184, freeram=86139719680, sharedram=117207040, bufferram=780677120, totalswap=728473600, freeswap=728473600, procs=3401, totalhigh=0, freehigh=0, mem_unit=1}) = 0
rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGQUIT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGQUIT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGTSTP, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGTSTP, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGTTIN, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGTTIN, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGTTOU, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGTTOU, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigaction(SIGQUIT, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
uname({sysname="Linux", nodename="X10DAi-00", ...}) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
stat("/home/werner", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat(".", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat("/home", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat("/home/werner", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
getpid()                                = 740126
getppid()                               = 740118
stat(".", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat("/home/werner/Public/repo/github.com/skywind3000/czmod.git/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/skywind3000/czmod.git/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/usr/local/lmod/lmod/libexec/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/.pyenv/plugins/pyenv-virtualenv/shims/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/.yarn/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/.cargo/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/.local/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/opt/apache-maven/apache-maven-3.8.1/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/mviereck/x11docker.git/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/opt/texlive/2021/bin/x86_64-linux/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/opt/openjdk/jdk-16.0.1/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/usr/local/nginx/sbin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/junegunn/fzf.git/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/kovidgoyal/calibre/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/txthinking/brook.git/cli/brook/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/.pyenv/shims/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/.pyenv/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/home/werner/Public/repo/github.com/golang/go.git/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/usr/local/sbin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/usr/local/bin/bash", 0x7ffe21050b00) = -1 ENOENT (No such file or directory)
stat("/usr/sbin/bash", 0x7ffe21050b00)  = -1 ENOENT (No such file or directory)
stat("/usr/bin/bash", {st_mode=S_IFREG|0755, st_size=1183448, ...}) = 0
stat("/usr/bin/bash", {st_mode=S_IFREG|0755, st_size=1183448, ...}) = 0
geteuid()                               = 1000
getegid()                               = 1000
getuid()                                = 1000
getgid()                                = 1000
access("/usr/bin/bash", X_OK)           = 0
stat("/usr/bin/bash", {st_mode=S_IFREG|0755, st_size=1183448, ...}) = 0
geteuid()                               = 1000
getegid()                               = 1000
getuid()                                = 1000
getgid()                                = 1000
access("/usr/bin/bash", R_OK)           = 0
stat("/usr/bin/bash", {st_mode=S_IFREG|0755, st_size=1183448, ...}) = 0
stat("/usr/bin/bash", {st_mode=S_IFREG|0755, st_size=1183448, ...}) = 0
geteuid()                               = 1000
getegid()                               = 1000
getuid()                                = 1000
getgid()                                = 1000
access("/usr/bin/bash", X_OK)           = 0
stat("/usr/bin/bash", {st_mode=S_IFREG|0755, st_size=1183448, ...}) = 0
geteuid()                               = 1000
getegid()                               = 1000
getuid()                                = 1000
getgid()                                = 1000
access("/usr/bin/bash", R_OK)           = 0
getpid()                                = 740126
getpgrp()                               = 740118
ioctl(2, TIOCGPGRP, [740118])           = 0
rt_sigaction(SIGCHLD, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, 8) = 0
prlimit64(0, RLIMIT_NPROC, NULL, {rlim_cur=514523, rlim_max=514523}) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
openat(AT_FDCWD, "/usr/local/lmod/lmod/init/bash", O_RDONLY) = 3
fstat(3, {st_mode=S_IFREG|0755, st_size=5716, ...}) = 0
read(3, "#!/bin/bash\n# -*- shell-script -"..., 5716) = 5716
close(3)                                = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
brk(0x563292fb1000)                     = 0x563292fb1000
openat(AT_FDCWD, "/dev/null", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 3
fcntl(2, F_GETFD)                       = 0
fcntl(2, F_DUPFD, 10)                   = 10
fcntl(2, F_GETFD)                       = 0
fcntl(10, F_SETFD, FD_CLOEXEC)          = 0
dup2(3, 2)                              = 2
close(3)                                = 0
openat(AT_FDCWD, "/usr/share/locale/en_CA/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale-langpack/en_CA/LC_MESSAGES/bash.mo", O_RDONLY) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=508, ...}) = 0
mmap(NULL, 508, PROT_READ, MAP_PRIVATE, 3, 0) = 0x14a80e04f000
close(3)                                = 0
openat(AT_FDCWD, "/usr/share/locale-langpack/en/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
fstat(2, {st_mode=S_IFCHR|0666, st_rdev=makedev(0x1, 0x3), ...}) = 0
ioctl(2, TCGETS, 0x7ffe2104fee0)        = -1 ENOTTY (Inappropriate ioctl for device)
write(2, "/usr/local/lmod/lmod/init/bash: "..., 65) = 65
dup2(10, 2)                             = 2
fcntl(10, F_GETFD)                      = 0x1 (flags FD_CLOEXEC)
close(10)                               = 0
faccessat(AT_FDCWD, "/usr/local/lmod/lmod/init/lmod_bash_completions", R_OK) = 0
openat(AT_FDCWD, "myclip.sh", O_RDONLY) = 3
stat("myclip.sh", {st_mode=S_IFREG|0664, st_size=123, ...}) = 0
ioctl(3, TCGETS, 0x7ffe21050e20)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = 0
read(3, "#! /bin/bash\nmyclip() {\nxclip -s"..., 80) = 80
lseek(3, 0, SEEK_SET)                   = 0
prlimit64(0, RLIMIT_NOFILE, NULL, {rlim_cur=1024, rlim_max=1024*1024}) = 0
fcntl(255, F_GETFD)                     = -1 EBADF (Bad file descriptor)
dup2(3, 255)                            = 255
close(3)                                = 0
fcntl(255, F_SETFD, FD_CLOEXEC)         = 0
fcntl(255, F_GETFL)                     = 0x8000 (flags O_RDONLY|O_LARGEFILE)
fstat(255, {st_mode=S_IFREG|0664, st_size=123, ...}) = 0
lseek(255, 0, SEEK_CUR)                 = 0
read(255, "#! /bin/bash\nmyclip() {\nxclip -s"..., 123) = 123
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
pipe([3, 4])                            = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, [INT CHLD], [], 8) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x14a80de01a10) = 740151
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigaction(SIGCHLD, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, 8) = 0
close(4)                                = 0
read(3, "x\n", 128)                     = 2
read(3, 0x7ffe21050680, 128)            = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=740151, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], WNOHANG, NULL) = 740151
wait4(-1, 0x7ffe2104ffd0, WNOHANG, NULL) = -1 ECHILD (No child processes)
rt_sigreturn({mask=[]})                 = 0
read(3, 
"", 128)                        = 0
close(3)                                = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigaction(SIGINT, {sa_handler=0x563292d1d480, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d1d480, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
pipe([3, 4])                            = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, [INT CHLD], [], 8) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x14a80de01a10) = 797652
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigaction(SIGCHLD, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, 8) = 0
close(4)                                = 0
read(3, "x\n", 128)                     = 2
read(3, 0x7ffe21050680, 128)            = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=797652, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], WNOHANG, NULL) = 797652
wait4(-1, 0x7ffe2104ffd0, WNOHANG, NULL) = -1 ECHILD (No child processes)
rt_sigreturn({mask=[]})                 = 0
read(3, "", 128)                        = 0
close(3)                                = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigaction(SIGINT, {sa_handler=0x563292d1d480, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d1d480, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
pipe([3, 4])                            = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, [INT CHLD], [], 8) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x14a80de01a10) = 797938
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigaction(SIGCHLD, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, 8) = 0
close(4)                                = 0
read(3, "x\n", 128)                     = 2
read(3, 0x7ffe21050680, 128)            = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=797938, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], WNOHANG, NULL) = 797938
wait4(-1, 0x7ffe2104ffd0, WNOHANG, NULL) = -1 ECHILD (No child processes)
rt_sigreturn({mask=[]})                 = 0
read(3, "", 128)                        = 0
close(3)                                = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigaction(SIGINT, {sa_handler=0x563292d1d480, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d1d480, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
pipe([3, 4])                            = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, [INT CHLD], [], 8) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x14a80de01a10) = 810458
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigaction(SIGCHLD, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, 8) = 0
close(4)                                = 0
read(3, "x\n", 128)                     = 2
read(3, 0x7ffe21050680, 128)            = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=810458, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], WNOHANG, NULL) = 810458
wait4(-1, 0x7ffe2104ffd0, WNOHANG, NULL) = -1 ECHILD (No child processes)
rt_sigreturn({mask=[]})                 = 0
read(3, "", 128)                        = 0
close(3)                                = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigaction(SIGINT, {sa_handler=0x563292d1d480, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d1d480, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x14a80de4a210}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
pipe([3, 4])                            = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, [INT CHLD], [], 8) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x14a80de01a10) = 810594
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigaction(SIGCHLD, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, {sa_handler=0x563292d20aa0, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x14a80de4a210}, 8) = 0
close(4)                                = 0
read(3, "x\n", 128)                     = 2
read(3, 0x7ffe21050680, 128)            = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=810594, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], WNOHANG, NULL) = 810594
wait4(-1, 0x7ffe2104ffd0, WNOHANG, NULL) = -1 ECHILD (No child processes)
rt_sigreturn({mask=[]})                 = 0
read(3,
```

**Comment by hongyi-zhao** [comment]
The following script works, see the discussion [here](https://groups.google.com/g/comp.unix.shell/c/QKgx7OH9yFk/m/Xk7Yl_ukCQAJ) for more detailed information:

```bash
$ cat myclip.sh
#! /bin/bash
myclip()
{
printf '%s\n' 'hello world' |
xclip -selection clipboard -in > /dev/null
printf '%s\n' x
sleep 1
return
}
set -x
while true; do
y="$(myclip)"
done

$ bash myclip.sh
+ true
++ myclip
++ printf '%s\n' 'hello world'
++ xclip -selection clipboard -in
++ printf '%s\n' x
++ sleep 1
++ return
+ y=x
+ true
++ myclip
++ printf '%s\n' 'hello world'
++ xclip -selection clipboard -in
++ printf '%s\n' x
++ sleep 1
++ return
+ y=x
+ true
++ myclip
++ printf '%s\n' 'hello world'
++ xclip -selection clipboard -in
^C
```

**Comment by mviereck** [comment]
I have compared the code.
The essential difference is to redirect stdout of `xclip` to `/dev/null`:
```
  xclip -selection clipboard -in <<< "hello world" > /dev/null
```
That is quite surprising.
I wonder if that is a bash or an xclip issue.

**Comment by hongyi-zhao** [comment]
Sometimes things like this require a deep understanding of Unix's underlying communication mechanisms. Perhaps the experts on `comp.unix.shell` can give more comprehensive analysis and explanation.

**Comment by gklanderman** [comment]
OP mis-understands how xclip works.

When you run
% xclip` -selection clipboard -in <<< "hello world"

you immediately get your prompt back, but xclip remains running in the background (it forked itself) and will keep running so long as it owns the selection.  When some other X11 application places data in the clipboard, only then it exits.  And at that point, as described by Helmut in detail above, the writing end of the pipe will close, and your $(myclip) will return.

Try this: run your script, and in another window:

% xclip -o -selection clipboard         
hello world
% xclip -o -selection clipboard         
hello world
% xclip -o -selection clipboard         
hello world

you can read from the clipboard multiple times, and the original xclip from your script keeps serving that (has not exited the $(myclip) yet).

but now do:

% xclip -selection clipbord -in <<< "replace"

and you'll see that the original xclip exits, myclip exits, and y is assigned.

Without knowing exactly what you're trying to achieve, it is hard to advise you, but you may want to look into the '-quiet' and '-loops' options.  

'-quiet' will cause xclip not to fork (stays in foreground), and '-loops' will cause it to exit after a given number or readers have read the data it posted to the clipboard.

**Comment by hongyi-zhao** [comment]
> % xclip -selection clipboard -in <<< "hello world"

This will put an additional `\n` into the clipboard, as described below:

```bash
$ man bash
**   Here Strings
       A variant of here documents, the format is:

              [n]<<<word

       The word undergoes tilde expansion, parameter and  variable  expansion,
       command  substitution,  arithmetic expansion, and quote removal.  Path‐
       name expansion and word splitting are not  performed.   The  result  is
       supplied as a single string, with a newline appended, to the command on
       its standard input (or file descriptor n if n is specified).
**
```

**Comment by gklanderman** [comment]
FWIW I went thru the open bugs and pull requests yesterday, and saw several that were proposing to have xclip -i close stderr before forking into the background.

So that would "fix" your example in the sense that myclip would no longer block, and instead return immediately, however, then you'll just be forking background xclip processes to replace themselves as fast as your shell can do so, which is almost certainly not what you want.  What are you actually trying to accomplish?

**Comment by mviereck** [comment]
Originally the issue came up in my project [x11docker](https://github.com/mviereck/x11docker).
The overall task to accomplish is to share the clipboard between two X servers.

A few days ago I tried a solution using xclip in a function where it reads from both X servers, compares the clips, sends the latest clip to both, and returns the latest clip. This was the point where I was stuck because the function was blocked by stdout of xclip and I did not understand why.

I don't think the behaviour of xclip should be changed as it is regular unix behaviour.  Redirecting stdout already solves the issue.

Meanwhile I have a new implementation that does not use a function.
FWIW, I have extracted my current implementation as a standalone script with `Xephyr` as a second X server:
```
#! /usr/bin/env bash

Hostdisplay=$DISPLAY
Hostxauthority=$XAUTHORITY
Newdisplay=:20
Xclientcookie=""

Xephyr $Newdisplay -ac &
sleep 1
DISPLAY=$Newdisplay xterm &

###############################
# share clipboard between X servers $Hostdisplay and $Newdisplay

X1auth[1]="DISPLAY=$Hostdisplay XAUTHORITY=$Hostxauthority"
X2auth[1]="DISPLAY=$Newdisplay XAUTHORITY=$Xclientcookie"
X1auth[2]="DISPLAY=$Newdisplay XAUTHORITY=$Xclientcookie"
X2auth[2]="DISPLAY=$Hostdisplay XAUTHORITY=$Hostxauthority"
Selection[1]=clipboard
Selection[2]=primary
OtherX[1]=2
OtherX[2]=1
Copyq[1]=':'
Copyq[2]=':'

while :; do
  for S in 1 2; do
    for X in 1 2; do
      export ${X1auth[$X]}
      Targets="$(xclip -out -selection ${Selection[$S]} -t TARGETS 2>/dev/null)"

      [ -n "$Targets" ] && {
        # check for image clip
        grep -q "image/png" <<< "$Targets" && Imagetarget="-t image/png" || Imagetarget=""

        # read content of ${Selection[$S]} of X server $X
        Clip[$X$S]="$(xclip -out -selection ${Selection[$S]} $Imagetarget | base64)"

        # check if ${Selection[$S]} of X server $X has changed
        [ -n "${Clip[$X$S]}" ] && [ "${Clipold[$X$S]}" != "${Clip[$X$S]}" ] && {
          Clipold[$X$S]="${Clip[$X$S]}"

          # send only to clipboard of other X server if it has different content. Important to keep highlighted text selection.
          [ "${Clip[${OtherX[$X]}$S]}" != "${Clip[$X$S]}" ] && {
            export ${X2auth[$X]}
            base64 -d <<< "${Clip[$X$S]}" | xclip -in -selection ${Selection[$S]} $Imagetarget
            #echo "SEND FROM $X (${Selection[$S]}): $(base64 -d <<< "${Clip[$X$S]}")" >&2
          }
        }
      }
    done
  done
  # sleep a bit to avoid high cpu usage
  sleep 0.3
done

```

################
A feature that would help me to improve this:
If `xclip -out` could behave similar to `xclip -in` in a way that it waits for new content before returning, I could write a less CPU expensive loop.
I think of a new option like `xclip -waitforchange` that returns once the clipboard content changes.
`xclip -in -quiet` already has this behaviour.
However, this leads to a special issue if it comes to `-selection primary`. The loop detects a new content with `xclip -out -selection primary` and comparing it with previous clipboard content.
But if I send this new content with `xclip -in -selection primary` to have a 'waiting xclip', the selected text looses it highlighting.
This might sound a bit confusing. I'll add a script that helps to reproduce and understand the issue.

**Comment by gklanderman** [comment]
> I don't think the behaviour of xclip should be changed as it is regular unix behaviour.
> Redirecting stdout already solves the issue as explained well above.

since xclip -i will not write to stdout, I think it actually does make sense to close stdout, and I think that was the conclusion drawn in several bugs, but I did not read them at all carefully.

> I think of a new option like xclip -waitforchange that returns once the clipboard
> content changes.

I implemented something similar (locally) today, so that the -wait option works with -o, but not in exactly the way you want:  what I implemented is that if no selection exists when you run xclip -o, it will wait up to the # of millisec specified by -wait, or indefinitely if -wait is negative.  It should not be too hard to change that to do what you want (just bypass the "if no current selection" check).

> this leads to a special issue if it comes to -selection primary

I think I see what you're getting at, but I can't think of a solution.

btw I just stumbled across this project the other day that sounds similar to what you're trying to do:

https://www.chiark.greenend.org.uk/~sgtatham/utils/xclipglue.html

have you seen it?

**Comment by mviereck** [comment]
> since xclip -i will not write to stdout, I think it actually does make sense to close stdout

At least with option `-filter` it writes to stdout.

> I implemented something similar (locally) today, so that the -wait option works with -o, but not in exactly the way you want: what I implemented is that if no selection exists when you run xclip -o, it will wait up to the # of millisec specified by -wait, or indefinitely if -wait is negative. It should not be too hard to change that to do what you want (just bypass the "if no current selection" check).

That sounds promising!
For my purpose `-wait` could be bound to `-out` as well; if I don't need the output, I could ignore it.

> btw I just stumbled across this project the other day that sounds similar to what you're trying to do:

Thank you for the hint! `xclipclue` looks promising and I wasn't aware of it. 
Unfortunately it has bugs. It does not sync the clipboards, instead it breaks the possibility to get context menus with right mouse click.

> I think I see what you're getting at, but I can't think of a solution.

Basically I only need a waiting xclip other than with `-in` to avoid expensive loops. I cannot use `-in` for this because it removes the highlighting of text to copy.

I've rewritten the code above a bit to split it into 4 loops.
Some sort of `xclip -out -waitforchange` could replace the `sleep` command. The script would react immediately on clipboard changes without a delay and it would not repeatedly compare the output of `xclip -out`  to no avail.

```
#! /usr/bin/env bash

Hostdisplay=$DISPLAY
Hostxauthority=$XAUTHORITY
Newdisplay=:20
Xclientcookie=""

Xephyr $Newdisplay -ac &
sleep 1
DISPLAY=$Newdisplay xterm &

#############################

clipexchange() {
  # share clipboard between two X servers
  # $1  credentials to access 1. X server
  # $2  credentials to access 2. X server
  # $3  selection to use (primary or clipboard)

  local Selection
  Selection="${3:-}"

  while :; do
    Targets="$(env ${1:-} xclip -out -selection $Selection -t TARGETS 2>/dev/null)"

    # Use xclip to wait for the clipboard to have some content
    [ -z "$Targets" ] && env ${1:-} xclip -selection $Selection -in -verbose <<< ""

    [ -n "$Targets" ] && {
      # check for image clip
      grep -q "image/png" <<< "$Targets" && Imagetarget="-t image/png" || Imagetarget=""

      # read content of $Selection of X server $1
      Clip="$(env ${1:-} xclip -selection $Selection -out $Imagetarget | base64)"

      # check if $Selection of X server $1 has changed
      [ -n "$Clip" ] && [ "$Clip" != "$Clipold" ] && {
        Clipold="$Clip"

        # send only to clipboard of other X server if it has different content. Important to keep highlighted text selection.
        Clipold2="$(env ${2:-} xclip -selection $Selection -out $Imagetarget | base64)"
        [ "$Clipold2" != "$Clip" ] && {
          #echo "SEND $Selection TO $2"
          base64 -d <<< "$Clip" | env ${2:-} xclip -selection $Selection -in $Imagetarget
        }
      }

      # wait a bit to avoid high CPU usage
      sleep 0.3

      # exit if one X server is not accessible
      env ${1:-} xclip -o >/dev/null || break
      env ${2:-} xclip -o >/dev/null || break
    }
  done
}

X1auth="DISPLAY=$Hostdisplay XAUTHORITY=$Hostxauthority"
X2auth="DISPLAY=$Newdisplay XAUTHORITY=$Xclientcookie"

clipexchange "$X1auth" "$X2auth" clipboard &
clipexchange "$X2auth" "$X1auth" clipboard &
clipexchange "$X1auth" "$X2auth" primary &
clipexchange "$X2auth" "$X1auth" primary &
```

**Comment by gklanderman** [comment]
sorry xclipglue didn't work, it really seems like it should do exactly what you want, and should ideally be more robust than a shell script.  you might reach out to the author, he last made a change ~1.5 years ago, and his email is at the bottom of this page:

https://www.chiark.greenend.org.uk/~sgtatham/utils/

re: your script, IIUC, you would be able to get rid of the sleep and the
 xclip -selection $Selection -in -verbose <<< ""
with the waitforchange flag?

I have a couple hoops with my employer before I can submit a pull request  for the wait feature I implemented.  Once I submit that, the additional option you want should be pretty easy to add.

**Comment by mviereck** [comment]
> re: your script, IIUC, you would be able to get rid of the sleep and the
xclip -selection $Selection -in -verbose <<< ""
with the waitforchange flag?

Yes, exactly.

> I have a couple hoops with my employer before I can submit a pull request for the wait feature I implemented. Once I submit that, the additional option you want should be pretty easy to add.

That sounds great! No need to hurry.

**Comment by hongyi-zhao** [comment]
The following is the relevant discussion Helmut Waitzmann sent me via private email on this issue. Because it is related to this issue, I also forward it here for reference:

> 
> The same "hanging" effect can be seen when using pipes in the
> commandline:  While in
> 
>    ( sleep -- 2 & date -- '+%Y-%m-%dT%T%z' ) | cat
>    date -- '+%Y-%m-%dT%T%z'
> 
> "cat" will wait for two seconds on the end‐of‐file condition, it
> will almost return immediately in
> 
>    ( sleep -- 2 > /dev/null & date -- '+%Y-%m-%dT%T%z' ) | cat
>    date -- '+%Y-%m-%dT%T%z'
> 
> 
> Kind regards
> 
> Helmut Waitzmann 
>

**Comment by gklanderman** [comment]
Sounds good Martin, this week is busy and I'm on vacation next week, so while I might get to it this week, it could take a few weeks.

**Comment by mviereck** [comment]
@gklanderman Just want to kindly ask if you still have this project in mind?

**Comment by gklanderman** [comment]
Hi @mviereck sorry I have had no time at all unfortunately.. I will try to make some progress in the next few weeks, ping me again if you don't hear anything..

**Comment by gklanderman** [comment]
Hi @mviereck sorry for the very long delay; not sure if you're still interested in this, but I no longer have employer hoops to jump through, and could probably make some time in the next month or so..

**Comment by quartzjer** [comment]
hi @gklanderman! I'm helping maintain xclip and all PRs are welcome anytime :+1:

**Comment by gklanderman** [comment]
Jer @quartzjer great to hear.. I have some changes I am using locally which I will try to clean up and send PRs for as a first step..

Then I can look at this issue and the related ones I mentioned above.

I see you're committing some changes lately.. do you have any plan for a release, and if so what's the timeline?  Looks like the last release was 8 or so years ago..
